home/rvagg/2021-02-exploration-report-small-filecoin-deals.md

```diff
@@ -0,0 +1,134 @@
+  * The most consistent number shown is power-of-two _minus_ the fr32 padding. (`2**Math.ceil(Math.log2(Math.ceil(10016091*(256/254))))/(256/254)/1024/1024` as `MiB`, or `15.875`, presented as `15.88 MiB` by Lotus - clear as mud).
+  * Aside: an implication of padding is that when packing data in for making efficient deals, you should get as close to the "power-of-two - fr32 padding" size as possible. The larger the data, the more waste you'll have. We had to do this for Filecoin Discover data to edge up as close as possible to 1 GiB pieces.
+* Found some miners
+  * Used the MinerX list (link to spreadsheet omitted)
+  * Ran `query-ask` against all of them (`for i in $(cat minerx-fellows.txt); do echo "querying $i"; lotus client query-ask $i >> minerx-fellows.txt.query-ask; done`); got 34 out of 36 responses.
+  * Filtered by minimum deal size, only 7 had miners accepting deals of 16 MiB or less
+  * Slate did some work on finding miners (not sure how up to date this is), collated @ https://sentinel.slate.host/api/mapped-static-global-miners and includes some miners that accept small deals too.
+* Ran `lotus client deal` to step through some deals interactively with miners I'd narrowed down. Got the following results for my file and submitted deals with most of these:
+  - f019104 ~0.0093312 FIL (0.000000018 FIL per epoch) ~$0.32
+  - f019041 ~0.0002592 FIL (0.0000000005 FIL per epoch) ~$0.01
+  - f066596 ~0.010368 FIL (0.00000002 FIL per epoch) ~$0.36
+  - f03134 ~8100 FIL (0.015625 FIL per epoch) ~$283,500 _(obviously I said `no` to this one)_
+  - f023565 ~0.0000000000005184 FIL (0.000000000000000001 FIL per epoch) $0
+  - f022352 ~0.0001458 FIL (0.00000000028125 FIL per epoch) ~$0.01 (0.005103)
+* Deal process is very slow and involves:
+  * Moving funds from the local wallet into the "StorageMarket" actor, to setup for a kind of escrow so the miner knows they're going to get paid
+  * The miner deciding whether to accept the funds. Miners set default deal parameters but can set additional configuration options and even insert an "deal acceptance script" to their deal flow to accept and reject deals based on all of the parameters. See https://docs.filecoin.io/mine/lotus/miner-configuration/#using-filters-for-fine-grained-storage-and-retrieval-deal-acceptance
+    * This can be slow too, I had deals stuck in `StorageDealCheckForAcceptance` for extended periods, I'm not sure what this is but maybe it's waiting for manual acceptance by the miner?
+  * Transferring the data via graphsync to the miner
+  * The miner setting up the deal with the chain (does this come before or after the next step?)
+  * The miner queueing the piece for sealing (can be slow, miners can choose to batch deals, which defer deals until enough have been selected, and can set maximum number of deals per sector too)
```

Comment:
Generally allowing 12~24h for queue+sealing is reasonable here. Anything under 12h is "WHOA!!! Really?!"